exchanges/gateio/gateio.go

```diff
@@ -424,169 +1013,2938 @@ func (g *Gateio) SendAuthenticatedHTTPRequest(ctx context.Context, ep exchange.U
+	}
+	contract.Delimiter = currency.UnderscoreDelimiter
+	params := url.Values{}
+	if leverage < 0 {
+		return nil, errInvalidLeverageValue
+	}
+	params.Set("leverage", strconv.FormatFloat(leverage, 'f', -1, 64))
+	if leverage == 0 && crossLeverageLimit > 0 {
+		params.Set("cross_leverage_limit", strconv.FormatFloat(crossLeverageLimit, 'f', -1, 64))
+	}
+	var response Position
+	return &response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, fmt.Sprintf(futuresDualModePositionLeverage, settle, contract.String()), params, nil, &response)
+}
+
+// UpdatePositionRiskLimitinDualMode update position risk limit in dual mode
+func (g *Gateio) UpdatePositionRiskLimitinDualMode(ctx context.Context, settle string, contract currency.Pair, riskLimit float64) ([]Position, error) {
+	if settle != settleBTC && settle != settleUSD && settle != settleUSDT {
+		return nil, errMissingSettleCurrency
+	}
+	if contract.Base.IsEmpty() || contract.Quote.IsEmpty() {
+		return nil, fmt.Errorf("%w, currency pair for contract must not be empty", errInvalidOrMissingContractParam)
+	}
+	params := url.Values{}
+	contract.Delimiter = currency.UnderscoreDelimiter
+	if riskLimit < 0 {
+		return nil, errInvalidRiskLimit
+	}
+	params.Set("risk_limit", strconv.FormatFloat(riskLimit, 'f', -1, 64))
+	var response []Position
+	return response, g.SendAuthenticatedHTTPRequest(ctx,
+		exchange.RestSpot, http.MethodGet,
+		fmt.Sprintf(futuresDualModePositionsRiskLimit, settle, contract.String()), params,
+		nil, &response)
+}
+
+// PlaceFuturesOrder creates futures order
+// Create a futures order
+// Creating futures orders requires size, which is number of contracts instead of currency amount. You can use quanto_multiplier in contract detail response to know how much currency 1 size contract represents
+// Zero-filled order cannot be retrieved 10 minutes after order cancellation. You will get a 404 not found for such orders
+// Set reduce_only to true can keep the position from changing side when reducing position size
+// In single position mode, to close a position, you need to set size to 0 and close to true
+// In dual position mode, to close one side position, you need to set auto_size side, reduce_only to true and size to 0
+func (g *Gateio) PlaceFuturesOrder(ctx context.Context, arg *OrderCreateParams) (*Order, error) {
+	if arg == nil {
+		return nil, errNilArgument
+	}
+	if arg.Contract.Base.IsEmpty() || arg.Contract.Quote.IsEmpty() {
+		return nil, fmt.Errorf("%w, currency pair for contract must not be empty", errInvalidOrMissingContractParam)
+	}
+	if arg.Size == 0 {
+		return nil, fmt.Errorf("%w, specify positive number to make a bid, and negative number to ask", errInvalidOrderSide)
+	}
+	if arg.TimeInForce != gtcTIF && arg.TimeInForce != iocTIF && arg.TimeInForce != pocTIF && arg.TimeInForce != focTIF {
+		return nil, errInvalidTimeInForce
+	}
+	if arg.Price > 0 && arg.TimeInForce == iocTIF {
+		arg.Price = 0
```

Comment:
I would actually opt to return an error. This obfuscates the issue that when in instant or cancel TIF it needs to be zero which should be set correctly before calling this function.